### PR TITLE
feat(cellguide): parallelize ontology tree state builder and canonical marker genes

### PR DESCRIPTION
**Pull request by atarashansky**
## Reason for Change

- This parallelizes a couple other parts of the pipeline, namely the ontology tree state builder for cell types and tissues, and the ASCTB table parsing.

## Changes

- add
- remove
- modify

## Testing steps

- Either list QA steps or reasoning you feel QA is unnecessary
- Describe how you made sure to know that your changes worked. Should allow someone else to go verify your code without in depth knowledge.
- "Unit tested only", "tested in rdev by a, b, c, verifying feature worked by... ", "manually ran pipeline locally with these results: ..."

### CODE REVIEW

File: backend/cellguide/pipeline/canonical_marker_genes/canonical_markers.py [hunk 1]
```diff
@@ -122,21 +124,27 @@ def _get_references(self, references: list[Reference], doi_to_citation: Dict[str
...
                 if doi not in doi_to_citation:
                     title = get_title_and_citation_from_doi(doi)
                     doi_to_citation[doi] = title
                 else:
                     title = doi_to_citation[doi]
-                refs.append(doi)
-                titles.append(title)
+                return doi, title
+
+        with ThreadPoolExecutor() as executor:
```
Comment (atarashansky) [inline backend/cellguide/pipeline/canonical_marker_genes/canonical_markers.py:137]: Using a ThreadPool to parallelize crossref fetches

File: backend/cellguide/pipeline/canonical_marker_genes/canonical_markers.py [hunk 2]
```diff
@@ -145,55 +153,18 @@ def get_processed_asctb_table_entries(self) -> Dict[str, ParsedAsctbTableEntry]:
...
         # DOI to citation mapping
-        doi_to_citation = {}
-
-        hashed_entries_seen = []
-        parsed_table_entries = []
-
-        for i, tissue in enumerate(self.asctb_data):
-            version = self.asctb_data[tissue]["metadata"]["version"]
-            logger.info(
-                f"Getting processed ASCTB table entry for {tissue} tissue ({version}) ({i+1}/{len(self.asctb_data)})"
```
Comment (atarashansky) [inline backend/cellguide/pipeline/canonical_marker_genes/canonical_markers.py:161]: I moved the stuff in this loop to a separate function that can be parallelized.

File: backend/cellguide/pipeline/ontology_tree/tree_builder.py [hunk 3]
```diff
@@ -277,52 +279,75 @@ def get_ontology_tree_state_per_celltype(self) -> Dict[str, OntologyTreeState]:
...
-                    end_node_i = end_node + "__" + str(i)
-                    path = self._depth_first_search_pathfinder(end_node_i)
-                    path = path if path else [end_node_i]
-                    all_paths.append(path)
+        return all_states_per_cell_type
 
-                    siblings.append(
-                        sum([self.all_children.get(parent, []) for parent in self.all_parents.get(end_node_i, [])], [])
-                    )
+    def _process_cell_type__parallel(self, celltype: str) -> OntologyTreeState:
```
Comment (atarashansky) [inline backend/cellguide/pipeline/ontology_tree/tree_builder.py:299]: this function is the outer loop for processing the cell types and is parallelized

File: backend/cellguide/pipeline/ontology_tree/tree_builder.py [hunk 4]
```diff
@@ -347,96 +372,117 @@ def get_ontology_tree_state_per_tissue(self) -> Dict[str, OntologyTreeState]:
...
-            df_rollup = rollup_across_cell_type_descendants(df, ignore_cols=["n_cells"])
-            df_rollup = df_rollup[df_rollup["n_cells_rollup"] > 0]
-
-            celltype_counts_in_tissue = dict(
-                zip(
-                    df_rollup["cell_type_ontology_term_id"],
-                    df_rollup[["n_cells", "n_cells_rollup"]].to_dict(orient="records"),
-                )
-            )
+    def _process_tissue__parallel(self, tissueId: str) -> OntologyTreeState:
```
Comment (atarashansky) [inline backend/cellguide/pipeline/ontology_tree/tree_builder.py:391]: this function contains the stuff in the outer loop for processing tissues and is parallelized

**Review by joyceyan — APPROVED** [review]

**Pull request merged**